indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java

```diff
@@ -1197,109 +1159,82 @@ private void startPendingTaskHandling()
             continue;
           }
-        }
 
-        String taskId = taskItem.getTaskId();
+          // Mark this worker as unassignable while task is being assigned
+          workers.compute(
+              workerToAssign.getWorker().getHost(), (key, entry) -> {
+                Preconditions.checkState(
+                    entry != null,
+                    "Expected selected worker[%s] to be available",
+                    entry.getWorker().getHost()
+                );
+                Preconditions.checkState(
+                    entry.getState() == WorkerHolder.State.READY,
+                    "Expected worker[%s] state to be READY, got [%s]",
+                    entry.getWorker().getHost(),
+                    entry.getState()
+                );
 
-        if (taskItem.getTask() == null) {
-          log.makeAlert("No Task obj found in TaskItem for taskID[%s]. Failed.", taskId).emit();
-          // taskComplete(..) must be called outside of statusLock, see comments on method.
-          taskComplete(
-              taskItem,
-              null,
-              TaskStatus.failure(
-                  taskId,
-                  "No payload found for this task. "
-                  + "See overlord logs and middleManager/indexer logs for more details."
-              )
+                entry.setState(WorkerHolder.State.PENDING_ASSIGN);
+                return entry;
+              }
           );
-          continue;
-        }
 
-        if (immutableWorker == null) {
-          throw new ISE("Unexpected state: null immutableWorker");
+          // Mark this task as pending worker assign
+          tasks.compute(
+              taskId,
+              (key, entry) -> {
+                Preconditions.checkState(entry != null, "Expected task[%s] to be in tasks set", taskId);
+                Preconditions.checkState(
+                    entry.getState() == HttpRemoteTaskRunnerWorkItem.State.PENDING,
+                    "Expected task[%s] state to be PENDING, got state[%s]",
+                    taskId,
+                    entry.getState()
+                );
+
+                entry.setState(HttpRemoteTaskRunnerWorkItem.State.PENDING_WORKER_ASSIGN);
+                return entry;
+              }
+          );
         }
 
+        final String workerHost = workerToAssign.getWorker().getHost();
         try {
-          // this will send HTTP request to worker for assigning task
-          if (!runTaskOnWorker(taskItem, immutableWorker.getWorker().getHost())) {
-            if (taskItem.getState() == HttpRemoteTaskRunnerWorkItem.State.PENDING_WORKER_ASSIGN) {
-              taskItem.revertStateFromPendingWorkerAssignToPending();
-            }
+          if (!runTaskOnWorker(taskId, workerHost)) {
+            log.warn("Failed to assign task[%s] to worker[%s]. Sending to back of queue", taskId, workerHost);
+            pendingTasks.put(taskItem.withFreshSequenceNumber());
```

Comment:
Elsewhere, we have used `pendingTasks.offer()`. If there isn't a specific reason to use one or the other, let's keep it consistent and use the same method in all the places.